src/core/fetchers/cdn_prioritizer.ts

```diff
@@ -0,0 +1,410 @@
+} from "../../utils/task_canceller";
+import SteeringManifestFetcher from "./steering_manifest";
+
+/**
+ * Class signaling the priority between multiple CDN available for any given
+ * resource.
+ *
+ * It might rely behind the hood on a fetched document giving priorities such as
+ * a Content Steering Manifest and also on issues that appeared with some given
+ * CDN in the [close] past.
+ *
+ * This class might perform requests and schedule timeouts by itself to keep its
+ * internal list of CDN priority up-to-date.
+ * When it is not needed anymore, you should call the `dispose` method to clear
+ * all resources.
+ *
+ * @class CdnPrioritizer
+ */
+export default class CdnPrioritizer extends EventEmitter<ICdnPrioritizerEvents> {
+  /**
+   * Metadata parsed from the last Content Steering Manifest loaded.
+   *
+   * `null` either if there's no such Manifest or if it is currently being
+   * loaded for the first time.
+   */
+  private _lastSteeringManifest : ISteeringManifest | null;
+
+  private _defaultCdnId : string | undefined;
+
+  /**
+   * Structure keeping a list of CDN currently downgraded.
+   * Downgraded CDN immediately have a lower priority than any non-downgraded
+   * CDN for a specific amount of time.
+   */
+  private _downgradedCdnList : {
+    /** Metadata of downgraded CDN, in no important order. */
+    metadata : ICdnMetadata[];
+    /**
+     * Timeout ID (to give to `clearTimeout`) of elements in the `metadata`
+     * array, for the element at the same index in the `metadata` array.
+     *
+     * This structure has been writted as an object of two arrays of the same
+     * length, instead of an array of objects, to simplify the usage of the
+     * `metadata` array which is used considerably more than the `timeouts`
+     * array.
+     */
+    timeouts : number[];
+  };
+
+  /**
+   * TaskCanceller allowing to abort the process of loading and refreshing the
+   * Content Steering Manifest.
+   * Set to `null` when no such process is pending.
+   */
+  private _steeringManifestUpdateCanceller : TaskCanceller | null;
+
+  private _readyState : ISharedReference<"not-ready" | "ready" | "disposed">;
+
+  /**
+   * @param {Object} manifest
+   * @param {Object|null} steeringManifestFetcher
+   * @param {Object} destroySignal
+   */
+  constructor(
+    manifest : Manifest,
+    steeringManifestFetcher : SteeringManifestFetcher | null,
+    destroySignal : CancellationSignal
+  ) {
+    super();
+    this._lastSteeringManifest = null;
+    this._downgradedCdnList = { metadata: [], timeouts: [] };
+    this._steeringManifestUpdateCanceller = null;
+    this._defaultCdnId = manifest.contentSteering?.defaultId;
+
+    let lastContentSteering = manifest.contentSteering;
+
+    manifest.addEventListener("manifestUpdate", () => {
+      const prevContentSteering = lastContentSteering;
+      lastContentSteering = manifest.contentSteering;
+      if (prevContentSteering === null) {
+        if (lastContentSteering !== null) {
```

Comment:
The naming `prevContentSteering` and `lastContentSteering` is a bit confusing for me, maybe `updatedContentSteering` or `newContentSteering`

Reply:
Yes I agree this is confusing. Maybe `lastContentSteering` should b renamed `currentContentSteering`? ("new" and "updated" make it seems that it is always new, whereas in the scope where it is declared, it is just the latest one).

Reply:
Done with `currentContentSteering`